File: results/dnd-rest-benchmark/lifecycle-runs/20260730T072747Z_pi_kimi-k2p7-code_go-stdlib/play_idempotent_events.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"strings"
)

// idempotentEventRequest binds the payload for creating an idempotent event.
type idempotentEventRequest struct {
	EventID string `json:"event_id"`
	Value   string `json:"value"`
}

// idempotentEventResponse is the immutable event shape returned by the create
// endpoint. The field order matches the exact JSON required by the stage
// contract.
type idempotentEventResponse struct {
	EventID        string `json:"event_id"`
	Value          string `json:"value"`
	Sequence       int    `json:"sequence"`
	IdempotencyKey string `json:"idempotency_key"`
}

// idempotentEventsListResponse is the ordered list returned by the read
// endpoint.
type idempotentEventsListResponse struct {
	Events []idempotentEventResponse `json:"events"`
}

// idempotentEventRow is the durable shape stored in the database.
type idempotentEventRow struct {
	CampaignID     string `json:"campaign_id"`
	Sequence       int    `json:"sequence"`
	EventID        string `json:"event_id"`
	Value          string `json:"value"`
	IdempotencyKey string `json:"idempotency_key"`
}

// nextIdempotentEventSequence returns the next monotonic sequence number for a
// campaign's idempotent event log. The caller must hold dbMu.
func nextIdempotentEventSequence(campaignID string) (int, error) {
	out, err := dbQuery(fmt.Sprintf("SELECT COALESCE(MAX(sequence), 0) + 1 AS next_seq FROM campaign_idempotent_events WHERE campaign_id=%s;", sq(campaignID)))
	if err != nil {
		return 0, err
	}
	var rows []struct {
		NextSeq int `json:"next_seq"`
	}
	if err := json.Unmarshal(out, &rows); err != nil {
		return 0, err
	}
	if len(rows) == 0 {
		return 1, nil
	}
	return rows[0].NextSeq, nil
}

// queryIdempotentEventByKey loads the stored event for a campaign and
// idempotency key, if any. The caller must hold dbMu.
func queryIdempotentEventByKey(campaignID, key string) (*idempotentEventRow, bool, error) {
	out, err := dbQuery(fmt.Sprintf("SELECT campaign_id, sequence, event_id, value, idempotency_key FROM campaign_idempotent_events WHERE campaign_id=%s AND idempotency_key=%s LIMIT 1;", sq(campaignID), sq(key)))
	if err != nil {
		return nil, false, err
	}
	var rows []idempotentEventRow
	if err := json.Unmarshal(out, &rows); err != nil {
		return nil, false, err
	}
	if len(rows) == 0 {
		return nil, false, nil
	}
	return &rows[0], true, nil
}

// queryIdempotentEvents loads all idempotent events for a campaign ordered by
// sequence. The caller must hold dbMu.
func queryIdempotentEvents(campaignID string) ([]idempotentEventRow, error) {
	out, err := dbQuery(fmt.Sprintf("SELECT campaign_id, sequence, event_id, value, idempotency_key FROM campaign_idempotent_events WHERE campaign_id=%s ORDER BY sequence;", sq(campaignID)))
	if err != nil {
		return nil, err
	}
	var rows []idempotentEventRow
	if err := json.Unmarshal(out, &rows); err != nil {
		return nil, err
	}
	if rows == nil {
		return []idempotentEventRow{}, nil
	}
	return rows, nil
}

// createIdempotentEventHandler creates a campaign-scoped idempotent event.
// Authenticated campaign members (including the owner) may create. Duplicate
// mutating requests with the same idempotency key have exactly one public
// effect.
func createIdempotentEventHandler(w http.ResponseWriter, r *http.Request) {
	dbMu.Lock()
	defer dbMu.Unlock()

	username, ok := bearerUsername(r)
	if !ok {
		writeError(w, http.StatusUnauthorized, "unauthorized")
		return
	}
	_, ok, err := loadUserByUsername(username)
	if err != nil {
		log.Printf("idempotent event auth query error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	if !ok {
		writeError(w, http.StatusForbidden, "forbidden")
		return
	}

	campaignID := r.PathValue("id")
	campaign, ok, err := queryPlayCampaign(campaignID)
	if err != nil {
		log.Printf("idempotent event campaign query error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	if !ok {
		writeError(w, http.StatusNotFound, "campaign not found")
		return
	}

	isMember := campaign.Owner == username
	if !isMember {
		isMember, err = isPlayCampaignMember(campaignID, username)
		if err != nil {
			log.Printf("idempotent event member query error: %v", err)
			writeError(w, http.StatusInternalServerError, "internal error")
			return
		}
	}
	if !isMember {
		writeError(w, http.StatusForbidden, "forbidden")
		return
	}

	idempotencyKey := strings.TrimSpace(r.Header.Get("Idempotency-Key"))
	if idempotencyKey == "" {
		writeError(w, http.StatusBadRequest, "invalid idempotency key")
		return
	}

	var req idempotentEventRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	if req.EventID == "" || req.Value == "" {
		writeError(w, http.StatusBadRequest, "invalid idempotent event")
		return
	}

	stored, exists, err := queryIdempotentEventByKey(campaignID, idempotencyKey)
	if err != nil {
		log.Printf("idempotent event key query error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	if exists {
		if stored.EventID == req.EventID && stored.Value == req.Value {
			writeJSON(w, http.StatusOK, idempotentEventResponse{
				EventID:        stored.EventID,
				Value:          stored.Value,
				Sequence:       stored.Sequence,
				IdempotencyKey: stored.IdempotencyKey,
			})
			return
		}
		writeError(w, http.StatusConflict, "idempotency conflict")
		return
	}

	exists, err = queryExists(fmt.Sprintf("SELECT 1 FROM campaign_idempotent_events WHERE campaign_id=%s AND event_id=%s LIMIT 1;", sq(campaignID), sq(req.EventID)))
	if err != nil {
		log.Printf("idempotent event duplicate event_id query error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	if exists {
		writeError(w, http.StatusConflict, "event_id already exists")
		return
	}

	sequence, err := nextIdempotentEventSequence(campaignID)
	if err != nil {
		log.Printf("idempotent event sequence query error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}

	if err := dbExec(fmt.Sprintf("INSERT INTO campaign_idempotent_events (campaign_id, sequence, event_id, value, idempotency_key) VALUES (%s, %d, %s, %s, %s);",
		sq(campaignID), sequence, sq(req.EventID), sq(req.Value), sq(idempotencyKey))); err != nil {
		log.Printf("idempotent event insert error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}

	writeJSON(w, http.StatusCreated, idempotentEventResponse{
		EventID:        req.EventID,
		Value:          req.Value,
		Sequence:       sequence,
		IdempotencyKey: idempotencyKey,
	})
}

// listIdempotentEventsHandler returns the ordered idempotent event log for a
// campaign. The campaign owner and members may read it.
func listIdempotentEventsHandler(w http.ResponseWriter, r *http.Request) {
	dbMu.Lock()
	defer dbMu.Unlock()

	username, ok := bearerUsername(r)
	if !ok {
		writeError(w, http.StatusUnauthorized, "unauthorized")
		return
	}
	_, ok, err := loadUserByUsername(username)
	if err != nil {
		log.Printf("idempotent events list auth query error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	if !ok {
		writeError(w, http.StatusForbidden, "forbidden")
		return
	}

	campaignID := r.PathValue("id")
	campaign, ok, err := queryPlayCampaign(campaignID)
	if err != nil {
		log.Printf("idempotent events list campaign query error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	if !ok {
		writeError(w, http.StatusNotFound, "campaign not found")
		return
	}

	isReader := campaign.Owner == username
	if !isReader {
		isReader, err = isPlayCampaignMember(campaignID, username)
		if err != nil {
			log.Printf("idempotent events list member query error: %v", err)
			writeError(w, http.StatusInternalServerError, "internal error")
			return
		}
	}
	if !isReader {
		writeError(w, http.StatusForbidden, "forbidden")
		return
	}

	rows, err := queryIdempotentEvents(campaignID)
	if err != nil {
		log.Printf("idempotent events list query error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}

	events := make([]idempotentEventResponse, 0, len(rows))
	for _, row := range rows {
		events = append(events, idempotentEventResponse{
			EventID:        row.EventID,
			Value:          row.Value,
			Sequence:       row.Sequence,
			IdempotencyKey: row.IdempotencyKey,
		})
	}

	writeJSON(w, http.StatusOK, idempotentEventsListResponse{Events: events})
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260731T051216Z_claude_claude-sonnet-5_php-stdlib/lib/play_idempotent_events.php
<?php
declare(strict_types=1);

// Routes: Play campaign idempotent events
// ---------------------------------------------------------------------------

function next_idempotent_event_sequence(PDO $db, string $campaignId): int {
    $stmt = $db->prepare('SELECT COALESCE(MAX(sequence), 0) AS max_seq FROM play_campaign_idempotent_events WHERE campaign_id = ?');
    $stmt->execute([$campaignId]);
    return (int)$stmt->fetch(PDO::FETCH_ASSOC)['max_seq'] + 1;
}

function idempotent_event_row_to_json(array $row): array {
    return [
        'event_id' => $row['event_id'],
        'value' => $row['value'],
        'sequence' => (int)$row['sequence'],
        'idempotency_key' => $row['idempotency_key'],
    ];
}

if ($method === 'POST' && preg_match('#^/v1/play/campaigns/([^/]+)/idempotent-events$#', $path, $pm)) {
    $campaignId = $pm[1];
    $actor = require_actor($db);

    $campaign = require_play_campaign($db, $campaignId);

    $isOwner = $actor['username'] === $campaign['owner'];
    if (!$isOwner && !is_campaign_member($db, $campaignId, $actor['username'])) {
        forbidden('must be a campaign member to create idempotent events');
    }

    $idempotencyKey = trim($_SERVER['HTTP_IDEMPOTENCY_KEY'] ?? '');
    if ($idempotencyKey === '') {
        bad_request('Idempotency-Key header is required');
    }

    $body = read_json_body();
    if ($body === null || !isset($body['event_id'], $body['value'])
        || !is_string($body['event_id']) || trim($body['event_id']) === ''
        || !is_string($body['value']) || trim($body['value']) === '') {
        bad_request('event_id and value are required nonempty strings');
    }
    $eventId = $body['event_id'];
    $value = $body['value'];

    $stmt = $db->prepare('SELECT * FROM play_campaign_idempotent_events WHERE campaign_id = ? AND idempotency_key = ?');
    $stmt->execute([$campaignId, $idempotencyKey]);
    $existingByKey = $stmt->fetch(PDO::FETCH_ASSOC);

    if ($existingByKey !== false) {
        if ($existingByKey['event_id'] === $eventId && $existingByKey['value'] === $value) {
            send_json(idempotent_event_row_to_json($existingByKey), 200);
        }
        conflict('idempotency key already used with a different event_id or value');
    }

    $stmt = $db->prepare('SELECT 1 FROM play_campaign_idempotent_events WHERE campaign_id = ? AND event_id = ?');
    $stmt->execute([$campaignId, $eventId]);
    if ($stmt->fetch(PDO::FETCH_ASSOC) !== false) {
        conflict('event_id already used in this campaign');
    }

    $sequence = next_idempotent_event_sequence($db, $campaignId);
    $stmt = $db->prepare('INSERT INTO play_campaign_idempotent_events (campaign_id, sequence, event_id, idempotency_key, value) VALUES (?, ?, ?, ?, ?)');
    $stmt->execute([$campaignId, $sequence, $eventId, $idempotencyKey, $value]);

    send_json([
        'event_id' => $eventId,
        'value' => $value,
        'sequence' => $sequence,
        'idempotency_key' => $idempotencyKey,
    ], 201);
}

if ($method === 'GET' && preg_match('#^/v1/play/campaigns/([^/]+)/idempotent-events$#', $path, $pm)) {
    $campaignId = $pm[1];
    $actor = require_actor($db);

    $campaign = require_play_campaign($db, $campaignId);

    $isOwner = $actor['username'] === $campaign['owner'];
    if (!$isOwner && !is_campaign_member($db, $campaignId, $actor['username'])) {
        forbidden('must be a campaign member to read idempotent events');
    }

    $stmt = $db->prepare('SELECT * FROM play_campaign_idempotent_events WHERE campaign_id = ? ORDER BY sequence ASC');
    $stmt->execute([$campaignId]);
    $rows = $stmt->fetchAll(PDO::FETCH_ASSOC);

    send_json([
        'events' => array_map('idempotent_event_row_to_json', $rows),
    ]);
}
